src/controller/OpenRobertaController.vala

```diff
@@ -0,0 +1,238 @@
+                        pin_dialog.close ();
+                        // remember selected server
+                        config.set_string ("Common", "SelectedServer",
+                            open_roberta_window.selected_server);
+                        try {
+                            config.save_to_file (CONFIG);
+                        } catch (FileError err) {
+                            warning ("FileError: %s", err.message);
+                        }
+                    } else {
+                        if (executing_user_code) {
+                            debug ("program done, switching to tty1");
+                            executing_user_code = false;
+                            switch_to_brickman_screen ();
+                        }
+                    }
+                }
+                if (message == "executing") {
+                    debug ("program starts, switching to tty2");
+                    executing_user_code = true;
+                    switch_to_program_screen ();
+                }
+            }
+        }
+
+        void on_server_edit () {
+            var  custom_server_address = open_roberta_window.custom_server.label.text;
+            if (custom_server_address == "custom server") {
+                custom_server_address = "";
+            }
+            var dialog = new InputDialog (
+                "Please enter server address", custom_server_address);
+            weak InputDialog weak_dialog = dialog;
+            dialog.responded.connect ((accepted) => {
+                if (!accepted) {
+                    return;
+                }
+                custom_server_address = weak_dialog.text_value;
+                open_roberta_window.custom_server.label.text = custom_server_address;
+                config.set_string ("Common", "CustomServer", custom_server_address);
+                try {
+                    config.save_to_file (CONFIG);
+                } catch (FileError err) {
+                    warning ("FileError: %s", err.message);
+                }
+                on_server_connect (open_roberta_window.custom_server.button);
+            });
+            dialog.show ();
+        }
+
+        void on_server_connect (Button button) {
+            var server = (button.child as Label).text;
+            if ( server == "" || server == "custom server") {
+                on_server_edit ();
+            } else {
+                try {
+                    open_roberta_window.selected_server = server;
+                    var code = service.connect ("http://" + server);
+                    var label = new Label (code) {
+                        margin_top = 12,
+                        font = BrickManagerWindow.big_font
+                    };
+                    pin_dialog = new MessageDialog.with_content ("Pairing code", label);
+                    pin_dialog.closed.connect (() => { pin_dialog = null; });
```

Comment:
`pin_dialog` should be a local variable, then you don't need the `connect` here. As it is, you have a reference cycle, which will cause a memory leak.